### PR TITLE
[WIP] Concept demonstration of ProceduralSystem

### PR DESCRIPTION
**Pull request by RussTedrake**
Here is one concept of writing and debugging procedural code in the systems framework.  I also think we should offer FiniteStateMachine isa LeafSystem and/or FinieStateMachine isa Diagram, but this, I think, it what is closest in spirit to some of our discussions.

For a quick glance at the proposed developer experience, I recommend you skip directly to reading `MyProceduralSystem::DoScript()`.

This example (of a quadrotor) was chosen to illustrate a high-rate control task + low-rate procedural code living in harmony.  

There are some very rough edges (it's been a while since I've thought hard about multithreading!), but I hope we can discuss the concept and its possibilities/limitations before we tear apart my exception vulnerability.

@sammy-tri @calderpg-tri @siyuanfeng-tri @sherm1 @jwnimmer-tri  -- wdyt?

---
This change is [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/15530)

### CODE REVIEW

**Comment by RussTedrake** [comment]
Let me be a little more precise.  I think exception vulnerability *should* be central to the discussion.  I just mean that we should separate whether it is possible to make this concept robust, as opposed to Russ' ability to achieve that status in a quick first draft.

**Comment by siyuanfeng-tri** [comment]
i took a quick pass, have two high level questions on top of my head:
1. what we currently have in anzu is also two threads, one for script, one for the robot bridge (which is the high rate thread here).  but these are not step locked. so doing time consuming planning in the script thread won't impact the control thread. i am not sure how to handle that case here.
2. it's not clear to me how the script thread here can interact with other procedural systems. the use case i had in mind was: i have one script controlling multiple independent robots async, and wait till they are all done.

**Review by RussTedrake — COMMENTED** [review]
Re: 1) I suppose that a procedural script could spawn more threads from/for the `DoScript()` as a part of the "perform arbitrary control flows".  But for long-running plans, etc, I would prefer to offer a separate `LeafSystem`-derived class (e.g. #11093) that makes that pattern efficient (and potentially more testable).

Re: 2) If the state of the multiple robots is coming in through an input port, then you can wait for it?  Much the same way as my `MoveTo` method below waits for the (single) robot to arrive within some threshold of the desired position, before moving on?

> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/15530#-:-MfvwmIs0D8PnaZ7GWgt:bvrts0u)* status: needs platform reviewer assigned, needs at least two assigned reviewers, labeled "do not merge"

**Review by amcastro-tri — COMMENTED** [review]
I've seen similar code in Anzu that when run purely in sim, it ends up not being deterministic. Is that a property we'd like to offer? (deterministic when purely in a sim with a single thread) or is this a intended for non-deterministic sims with sim and controller in different threads?

> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/15530#-:-MgBi0zc48vPHCLTyehZ:btuhlyq)* status: needs platform reviewer assigned, needs at least two assigned reviewers, labeled "do not merge"

**Review by RussTedrake — COMMENTED** [review]
This *will* be deterministic in sim if implemented correctly (unless, of course, the user code does something non-deterministic).  The second thread stays in lock step with the main simulation thread.  I tried to describe that in the code documentation, let me know if there is a part that should be more clear.

> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/15530#-:-MgDbae05dmoPTwg9yGt:blcc7zs)* status: needs platform reviewer assigned, needs at least two assigned reviewers, labeled "do not merge"

**Comment by jwnimmer-tri** [comment]
I would probably approach this differently, to make it both simpler and more powerful.

(1) As a minor implementation point, instead of a thread I would use async and future.  That is a better model of a long-running function call, and the safety / ergonomics will just be smoother overall.  Especially, in the case where we might want to have multiple concurrent function calls dispatching from within a single system, it should be much easier to manage.  However, for the purposes of understanding how a System should express concurrency, parallelism, and undeclared procedural state, async vs thread should be basically the same idea, so we can set it aside for now.

(2) I think we can raise our level of abstraction beyond a raw mutex, condition variables, and lockstep state updates.  As I understand the code here, the semantics you have for the quadrotor planner system are:

- (a) declare a periodic discrete update event;
- (b) within the planner system dynamics:
  - when the simulator (for example) asks for xd⁺, give the pre-update context to the worker function and wait on its reply;
    - once the worker function replies with xd⁺, give it back to the simulator.
- (c) within the worker function:
  - initialize (we should probably give it the current context for this purpose, though we don't appear to do so yet)
  - wait on the system to give us an updated context;
  - once the system gives us a new context, compute and reply with the new plan;
  - (... repeat / loop the above two steps as necessary ...)

Now, take that formulation and imagine that the information-passing was not via pointers to a (const) context or (mutable) plan with manual locks, but rather via a thread-safe queue.  A FIFO queue with an upper-bounded size where one thread pushes and another thread pops is a common idiom, and it's easy to make it block (sleep) the pusher when full, or block (sleep) the puller when empty.  With a max size of 1, we have the same semantics as you do here, just with a different implementation.  (A max-size-1 threadsafe queue is often called a "mailbox", as a term of art.)

Even with only that change, I'll bet the implementation within DoScript should tape out much more clearly.  Better, though, now that we've phrased it that way, we can easily turn the knob of what kind of _concurrency_ we are trying to model, and therefore how much _parallelism_ we can achieve.

For example, imagine a delay-line on the output of the planner -- that the state update at x[n+1] is a function of the context at x[n], one timestep earlier.  That's just a change to how the FIFO management within the planner system dynamics chooses to fill & drain the worker's queue -- it initially pushes two contexts before blocking on the first reply, and then push/pull pairwise after that.  It's still fully deterministic, but now you have scheduling parallelism instead of tick-tock scheduling.

Similar adjustments can be made to the "landing time" of plan updates -- depending the nature of the requests to plan, we can schedule its result to land back into the context at an arbitrary time into the future, not a fixed schedule.  Similarly, we can dispatch to multiple workers concurrently, and have some policy about which one will win the game to update the plan.  Or, the worker can slurp up multiple updates (or on-the-fly updates) while it plans off the input FIFO, to change what it places onto its output FIFO.

All of those can retain determinism (or not), while admitting different choices for concurrency and thus parallelism (or not).  But they all start with re-framing the worker as a long-running function that uses FIFO input queue and FIFO output queue, instead of a pair of pointers.

The key abstraction is message passing.  An important detail here is that, in a "procedural system", the worker is allowed to have memory (and so, we have #4330).  If we restrict the worker to be memoryless, though, now we are back to all state beind declared (and so, we can rewind / replay / resume).  And our choices for the dynamics model are exactly about how much or how little we specify the message passing to enforce rendezvous points for synchronization, such that the admitted message sequences are either fully determined, or at least bounded within some delay or scheduling regime.

In short, I think reformulating our system models with those concepts (and implementation tactics) in mind, will lead to a result that is easier to implement, reason about, and flex to meet to our needs.

**Comment by RussTedrake** [comment]
Thanks to those of you that provided feedback.  I will re-open this when it's ready to move forward.
